### PR TITLE
REGR: fix regression in scalar setitem with setting a length-1 array-like

### LINKED ISSUE #46268: REGR: ValueError: setting an array element with a sequence. when assigning an array.array using .loc
### Pandas version checks

- [X] I have checked that this issue has not already been reported.

- [X] I have confirmed this bug exists on the [latest version](https://pandas.pydata.org/docs/whatsnew/index.html) of pandas.

- [x] I have confirmed this bug exists on the main branch of pandas.

### Reproducible Example

```python
import pandas as pd
import numpy as np
from array import array

df = pd.DataFrame({"A": [2,4,6,8]})

cenlon = array('d')
cenlon.append(45.345343)

# Problem
df.loc[0, 'CenLon'] = cenlon

# Solution 1
# df.loc[0, 'CenLon'] = np.array(cenlon)

# Solution 2
# df['CenLon'] = ''
# df.loc[0, 'CenLon'] = cenlon
```

### Issue Description

If the Example above is running in a jupyter notebook, the notebook gets stuck and no error message is shown. The error message only appears in the console where the jupyter notebook is running. From the notebook inside it looks like the cell is still computing. The Example also includes two solutions to it, but a decent error message still would be nice.

### Expected Behavior

It would be great if the error message would be displayed directly in the notebook or the provided datatype should be supported again.
As background info: this datatype you will get when working with geopandas (https://geopandas.org/en/stable/index.html) and shaply (https://shapely.readthedocs.io/en/stable/manual.html) and you want to extract the centre coordinates of a geometry. Further, I think this was supported in a previous pandas version as the error just appeared on an older notebook (if wanted I could try to find out up to which version or you decide it is not supported anymore then I will let shaply and geopandas know).

### Installed Versions

<details>
INSTALLED VERSIONS
------------------
commit           : 06d230151e6f18fdb8139d09abf539867a8cd481
python           : 3.9.10.final.0
python-bits      : 64
OS               : Linux
OS-release       : 5.14.0-1024-oem
Version          : #26-Ubuntu SMP Thu Feb 17 14:35:50 UTC 2022
machine          : x86_64
processor        : x86_64
byteorder        : little
LC_ALL           : None
LANG             : en_US.UTF-8
LOCALE           : en_US.UTF-8

pandas           : 1.4.1
numpy            : 1.22.3
pytz             : 2021.3
dateutil         : 2.8.2
pip              : 22.0.4
setuptools       : 59.8.0
Cython           : 0.29.28
pytest           : 7.0.1
hypothesis       : None
sphinx           : None
blosc            : None
feather          : None
xlsxwriter       : None
lxml.etree       : None
html5lib         : None
pymysql          : None
psycopg2         : None
jinja2           : 3.0.3
IPython          : 8.1.1
pandas_datareader: None
bs4              : None
bottleneck       : None
fastparquet      : None
fsspec           : 2022.02.0
gcsfs            : None
matplotlib       : 3.5.1
numba            : None
numexpr          : 2.8.0
odfpy            : None
openpyxl         : None
pandas_gbq       : None
pyarrow          : None
pyreadstat       : None
pyxlsb           : None
s3fs             : None
scipy            : 1.8.0
sqlalchemy       : None
tables           : 3.7.0
tabulate         : None
xarray           : 2022.3.0
xlrd             : None
xlwt             : None
zstandard        : None

</details>

### PR DESCRIPTION
**Pull request by jorisvandenbossche**
- [x] closes #46268
- [x] [Tests added and passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#writing-tests) if fixing a bug or adding a new feature
- [ ] All [code checks passed](https://pandas.pydata.org/pandas-docs/dev/development/contributing_codebase.html#pre-commit).
- [x] Added an entry in the latest `doc/source/whatsnew/vX.X.X.rst` file if fixing a bug or adding a new feature.

While testing the various related cases, I noticed that we don't have very consistent behaviour always, see the TODO notes in the tests. 
In pandas 1.3, we always unpacked an array-like when setting, also for object dtype. This means that for object dtype, it unpacked it for len-1, and raised an error for longer arrays:

```
In [31]: df = pd.DataFrame({"A": [1, 2, 3], "B": np.array([1, 'a', 'b'], dtype=object)})

In [32]: df.loc[0, "B"] = np.array([0])

In [33]: df
Out[33]: 
   A  B
0  1  0
1  2  a
2  3  b

In [34]: df.loc[0, "B"] = np.array([0, 1])
...
ValueError: Must have equal len keys and value when setting with an iterable
```

With the change in https://github.com/pandas-dev/pandas/pull/42780, in pandas 1.4 we started to preserve the nested object in case of len-1, but still raising for longer arrays (with object dtype column):

```
In [24]: df = pd.DataFrame({"A": [1, 2, 3], "B": np.array([1, 'a', 'b'], dtype=object)})

In [25]: df.loc[0, "B"] = np.array([0])

In [26]: df
Out[26]: 
   A    B
0  1  [0]   # <-----
1  2    a
2  3    b

In [27]: df.loc[1, "B"] = np.array([0, 1])
...
ValueError: Must have equal len keys and value when setting with an iterable
```

This is a bit inconsistent I would say, and I think we should probably long term allow setting nested data of any length _in case of object dtype column_ (the first TODO note).

Because of the above change (no longer unpacking for len-1 array), this started to error for _numeric_ columns (i.e. the actual regression reported in #46268). 

What I did in this PR is limiting the change of https://github.com/pandas-dev/pandas/pull/42780 to only object dtype, so we keep the new behaviour of preserving the nested object when setting in case object dtype, while fixing the regression for numeric data. 
(strictly speaking, we could also say that it is a regression for the object dtype case, but there I think it was actually an improvement in behaviour, and since this was already changed in 1.3.5, and nobody reported it, it seems better to keep that now)

In addition, while testing the same for Series, I noticed that for Series setitem, we only do this unpacking of len-1 array-like values for actual numpy arrays, and not for setting array-likes in general (the second TODO note).

TODO: open new issues to keep track of those TODOs and update those comments

### CODE REVIEW

File: pandas/tests/indexing/test_indexing.py [hunk 1]
```diff
@@ -1019,3 +1020,95 @@ def test_ser_list_indexer_exceeds_dimensions(indexer_li):
...
+            "Must have equal len keys and value",
+            "setting an array element with a sequence",
+        ]
+    )
+    with pytest.raises(ValueError, match=msg):
+        df.loc[0, "B"] = value
+
+    # TODO For object dtype this happens as well, but should we rather preserve
+    # the nested data and set as such?
+    df = DataFrame({"A": [1, 2, 3], "B": np.array([1, "a", "b"], dtype=object)})
```
Comment (phofl) [inline pandas/tests/indexing/test_indexing.py:1042]: I think this sound good

File: pandas/tests/indexing/test_indexing.py [hunk 2]
```diff
@@ -1019,3 +1020,95 @@ def test_ser_list_indexer_exceeds_dimensions(indexer_li):
...
+    # https://github.com/pandas-dev/pandas/issues/46268
+
+    # For numeric data, assigning length-1 array to scalar position gets unpacked
+    df = DataFrame({"A": [1, 2, 3]})
+    df.loc[0, "B"] = value
+    expected = DataFrame({"A": [1, 2, 3], "B": [0.0, np.nan, np.nan]})
+    tm.assert_frame_equal(df, expected)
+
+    # but for object dtype we preserve the nested data
+    df = DataFrame({"A": [1, 2, 3], "B": np.array([1, "a", "b"], dtype=object)})
```
Comment (phofl) [inline pandas/tests/indexing/test_indexing.py:1082]: If we do the above, this is correct I think

File: pandas/tests/indexing/test_indexing.py [hunk 3]
```diff
@@ -1019,3 +1020,95 @@ def test_ser_list_indexer_exceeds_dimensions(indexer_li):
...
+        assert df.loc[0, "B"] == value
+
+
+@pytest.mark.parametrize(
+    "value", [(0.0,), [0.0], np.array([0.0]), array.array("d", [0.0])]
+)
+def test_scalar_setitem_series_with_nested_value_length1(value, indexer_sli):
+    # For numeric data, assigning length-1 array to scalar position gets unpacked
+    # TODO this only happens in case of ndarray, should we make this consistent
+    # for all list-likes? (as happens for DataFrame.(i)loc, see test above)
```
Comment (phofl) [inline pandas/tests/indexing/test_indexing.py:1096]: Yes, I think this should unpack for everything. The DataFrame case already does

**Review by phofl — COMMENTED** [review]
Some comments

**Comment by jorisvandenbossche** [comment]
@phofl just to be clear: those TODO comments I added are interesting points (and I should open new issues for those), but fixing those are out of scope of this focused regression fix for 1.4.4 (your feedback on those questions was very welcome! just to set expectations that I don't plan to "fix" those todos here ;)). See also the longer explanation in the top post.

**Comment by phofl** [comment]
Yep understood this, no worries. Just wanted to coment in code to avoid misunderstandings with the references

**Review by phofl — APPROVED** [review]
I think this is good to merge

**Review by mroeschke — APPROVED** [review]

**Comment by mroeschke** [comment]
Thanks @jorisvandenbossche (failures unrelated)

**Pull request merged**